File: backend/app/schemas/traffic.py
"""Schemas for Current Traffic API."""

from typing import List, Optional
from pydantic import BaseModel, Field
from backend.app.schemas.common import PaginationMeta, ProvenanceEnum


class TrafficRecord(BaseModel):
    """Traffic intelligence record for a road segment."""
    timestamp: str
    segment_id: str
    source_node: Optional[str] = None
    target_node: Optional[str] = None
    
    # Observed
    speed_kmh: float
    flow_vph: float
    occupancy_pct: float
    queue_length_veh: float
    delay_min: float

    # Model-Derived
    congestion_score: float
    congestion_state: str
    is_anomaly: int
    anomaly_type: str
    confidence: float
    temporal_status: str
    roadwork_context: str
    evidence_reason: str
    provenance: str = ProvenanceEnum.DERIVED.value


class TrafficListResponse(BaseModel):
    """Response containing list of traffic records and pagination metadata."""
    meta: PaginationMeta
    data: List[TrafficRecord]

File: frontend/types/traffic.ts
/**
 * Traffic intelligence types matching FastAPI /api/traffic/current response.
 */

import { ProvenanceType, PaginationMeta } from "./api";

export type CongestionState = "NORMAL" | "WATCH" | "CONGESTED" | "SEVERE";

export interface TrafficRecord {
  timestamp: string;
  segment_id: string;
  source_node?: string | null;
  target_node?: string | null;

  // Observed metrics
  speed_kmh: number;
  flow_vph: number;
  occupancy_pct: number;
  queue_length_veh: number;
  delay_min: number;

  // Model-derived metrics
  congestion_score: number;
  congestion_state: CongestionState | string;
  is_anomaly: number;
  anomaly_type: string;
  confidence: number;
  temporal_status: string;
  roadwork_context: string;
  evidence_reason: string;
  provenance: ProvenanceType | string;
}

export interface TrafficListResponse {
  meta: PaginationMeta;
  data: TrafficRecord[];
}

export interface TrafficQueryParams {
  limit?: number;
  offset?: number;
  segment_id?: string;
  congestion_state?: string;
}
